### PR TITLE
feat: indexing of token system stats

### PR DESCRIPTION
**Pull request by janb87**
## Summary

This PR implements comprehensive indexing of token system statistics, enabling real-time tracking of total system value across all tokens in a registry. The implementation includes persistent state management, timeseries data collection, and GraphQL query capabilities.

### 🔄 System Statistics Tracking
- **SystemStatsState**: Persistent entity tracking current total system value
- **SystemStatsData**: Timeseries entity for historical system statistics
- Automatic recalculation on token supply or price changes

### 📊 Real-time Updates
- Supply change handlers (mint/burn operations)
- Price change handlers (claim updates)
- Automatic system total value recalculation

### 🔗 Enhanced Token Factory Registry
- Associates each registry with its parent system
- Updated GraphQL schema and dapp-kit types
- Improved system-to-registry relationships

### 🛠 Utilities & Infrastructure
- `toBigDecimal` utility for decimal calculations
- Enhanced claim scheme utilities with tuple encoding
- Docker Compose health checks for Blockscout

## GraphQL Queries

New queries available:
- `systemStatsState(id: ID\!)` - Get current system statistics
- `systemStatsStates()` - List all system statistics with filtering

## Testing

- Comprehensive unit tests for system-stats indexing logic
- Validation of supply and price change scenarios
- Edge case handling and error conditions

## Technical Implementation

- **Handlers**: Updated mint, burn, and claim handlers to trigger system stats updates
- **Schema**: Extended GraphQL schema with new entities and relationships  
- **Types**: Generated TypeScript types for frontend integration
- **Utilities**: Refactored claim decoding and decimal handling

This enables the dapp to display real-time system-wide metrics and historical trends for token ecosystems.

🤖 Generated with [Claude Code](https://claude.ai/code)

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: kit/contracts/scripts/hardhat/utils/claim-scheme-utils.ts [hunk 1]
```diff
@@ -70,5 +70,14 @@ export function encodeClaimData(
     throw new Error(`Unknown claim topic: ${topicId}`);
   }
 
-  return encodeAbiParameters(parseAbiParameters(signature), values);
+  const abiParams = parseAbiParameters(signature);
+
+  // Need to wrap it in a tuple so it can be decoded by The Graph
+  // To get the "0x...20..." format (like The Graph's ethereum.decode expects for a tuple)
+  const encodedViemTuple = encodeAbiParameters(
```
Comment (janb87) [inline kit/contracts/scripts/hardhat/utils/claim-scheme-utils.ts:77]: See Gemini

You've hit on the core of the issue with the `decodeAbiParameters` vs. `ethereum.decode` behavior\!

You're right, in Viem:

```typescript
// Viem
const decoded = decodeAbiParameters(
  [
    { type: "string", name: "class" },
    { type: "string", name: "category" },
  ],
  data // "0x...40..." data
);
```

And in The Graph (AssemblyScript mapping):

```typescript
// The Graph (AssemblyScript)
let decoded = ethereum.decode("(string,string)", data); // data is the "0x...20..." data
```

**The fundamental difference is in how each library handles the "top-level" encoding of a tuple, especially when it's the *only* thing being encoded or decoded.**

Let's re-emphasize the ABI encoding rules for tuples:

1.  **Tuple Header:** If a tuple is a top-level element (e.g., a function return value, or a single parameter being encoded), its encoding typically starts with a series of 32-byte words, each representing the *offset* to the actual data of the corresponding element within the tuple.
2.  **Element Data:** After these offsets, the actual encoded data for each element (prefixed by its own length if it's a dynamic type) follows.

**How Viem (`decodeAbiParameters`) and The Graph (`ethereum.decode`) Interpret This:**

  * **Viem's `decodeAbiParameters` (and `encodeAbiParameters`):**
    When you provide an array of `AbiParameter` objects directly to `decodeAbiParameters`, Viem assumes you are decoding a sequence of parameters as they would appear if they were *multiple independent arguments* to a function, or a single tuple where the *offsets for all elements* are laid out immediately.

    The `0x...40...` data for `["Class A", "Category A"]` is a standard way to encode `(string, string)` if you consider it as two separate parameters being encoded sequentially, or a tuple where the initial block contains offsets for all its members.

      * The first `0x40` (64 decimal) means the *first actual dynamic data part* (the length of "Class A") starts at byte 64. This makes sense: 32 bytes for the offset of "Class A" + 32 bytes for the offset of "Category A" = 64 bytes of initial fixed data.

  * **The Graph's `ethereum.decode` (and `ethereum.encode`):**
    The `ethereum.decode` and `ethereum.encode` functions in AssemblyScript are designed to be extremely faithful to how Solidity itself encodes and decodes complex types, especially single function arguments or return values that are tuples.

    When Solidity returns a `(string, string)` tuple, or you pass a `(string, string)` as a single argument, the encoding often looks like the `0x...20...` data.

      * The first `0x20` (32 decimal) signifies that the data for the *first element of the tuple* starts 32 bytes into the encoding. In essence, the first 32 bytes *are* the offset for the first string. The subsequent `0x40` is then the offset for the second string *relative to the start of the tuple's data section*.

**In essence:**

  * **`0x...40...` (Viem's `decodeAbiParameters` preference):** This represents an encoding where the list of parameters is implicitly a "tuple" of its own, and the first 32-byte word is the offset to the *first dynamic data block* in that sequence of parameters.
  * **`0x...20...` (The Graph's `ethereum.decode` preference):** This represents an encoding of a *single tuple type*, where the first 32-byte word is the offset to the *first element's data within that tuple*.

**Why the difference matters and how to handle it:**

The `decodeAbiParameters` function in Viem is more flexible and can often correctly parse both. It understands that when given a list of `AbiParameter` objects, it should try to consume the offsets correctly.

The `ethereum.decode` in The Graph is often more strict about the top-level structure, expecting the format that Solidity itself would produce for a single tuple type.

**Your current tests show that `decodeAbiParameters` in Viem correctly handles *both* of these encoded forms.** This is a testament to Viem's robustness.

**What you need to keep in mind:**

  * **If you are generating data in Viem to be consumed by The Graph:** You'll want to ensure Viem produces the `0x...20...` format. You can achieve this by explicitly encoding a `tuple` type.

    ```typescript
    import { encodeAbiParameters, Hex } from 'viem';

    const args = ['Class A', 'Category A'];
    const types = [{ type: 'string' }, { type: 'string' }];

    // To get the "0x...40..." format (like decodeAbiParameters expects for direct params)
    const encodedViemParams = encodeAbiParameters(types, args) as Hex;
    console.log("Viem (direct parameters):", encodedViemParams); // Will produce the 0x...40... format

    // To get the "0x...20..." format (like The Graph's ethereum.decode expects for a tuple)
    const encodedViemTuple = encodeAbiParameters(
      [{ type: 'tuple', components: types }], // Explicitly encode as a single tuple
      [args] // Wrap the args in an array because the main type is now a single tuple
    ) as Hex;
    console.log("Viem (explicit tuple):", encodedViemTuple); // Will produce the 0x...20... format
    ```

    You'll notice `encodedViemTuple` will match your `dataTheGraph`.

  * **If you are consuming data from The Graph in Viem:** As your successful tests already show, `decodeAbiParameters` (with the simple `{ type: "string" }, { type: "string" }` array) is capable of decoding The Graph's `0x...20...` tuple encoding. This is the simplest and most effective approach.

**Conclusion:**

The difference in encoded data is a subtle but standard variation in ABI encoding based on whether something is interpreted as a sequence of independent parameters or a single, top-level tuple.

**Viem's `decodeAbiParameters` is flexible enough to correctly decode both formats you provided, which is great\!** You don't necessarily need `decodeFunctionResult` for this specific case if you're dealing with raw ABI-encoded tuples. Your current test setup where `decodeAbiParameters` handles `dataTheGraph` is the correct and most straightforward way to work with it.
